crates/libs/core/src/dynamic_cast.rs

```diff
@@ -0,0 +1,143 @@
+//! Provides support for dynamic casts from a COM interface ([`IDynamicCast`]) to a Rust
+//! implementation object.
+//!
+//! # Implementation notes
+//!
+//! This implementation manually implements the [`IDynamicCast`] type and its associated
+//! [`IDynamicCast_Vtbl`] and [`IDynamicCast_Impl`] items, instead of using the `#[implement]`
+//! macro. This is necessary for safely encapsulating unsafety at the interface boundary.
+//!
+//! It would be possible for `IDynamicCast_Impl` to define a method `fn as_any(&self) -> &dyn Any`.
+//! However, such an implementation would only have access to the "inner" type. We want access to
+//! the "outer" type. We solve this by defining the sole implementation of [`IDynamicCast`] within
+//! this crate, and having its implementation directly cast from the COM interface pointer to the
+//! _outer_ object pointer (the `MyApp_Impl` type). This allows us to increase the reference count
+//! and to synthesize new [`ComObject`] instances.
+//!
+//! The manual implementation of [`IDynamicCast`] is also needed because the implementation defines
+//! a method that uses the `Rust` calling convention, rather than `system`. This is necessary
+//! because the method must use Rust types (`&dyn Any`).
+
+use crate::imp::E_NOINTERFACE;
+use crate::*;
+use core::any::Any;
+use core::ffi::c_void;
+
+/// An interface that allows for dynamic casting, for all Rust types that implement [`core::any::Any`].
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct IDynamicCast(IUnknown);
+
+unsafe impl Interface for IDynamicCast {
+    type Vtable = IDynamicCast_Vtbl;
+    const IID: GUID = GUID::from_u128(0xae49d5cb_143f_431c_874c_2729336e4eca);
+}
```

Comment:
```suggestion
imp::define_interface!(IDynamicCast, IDynamicCast_Vtbl, 0xae49d5cb_143f_431c_874c_2729336e4eca);
```

Reply:
Resolved, because this code is now deleted.